src/main/java/org/testng/internal/thread/graph/GraphThreadPoolExecutor.java

```diff
@@ -127,35 +127,42 @@ private void mapNodeToWorker(List<IWorker<T>> runnables, List<T> freeNodes) {
     if (!RuntimeBehavior.enforceThreadAffinity()) {
       return;
     }
-    int index = 0;
     for (IWorker<T> runnable : runnables) {
-      T freeNode = freeNodes.get(index++);
-      IWorker<T> w = mapping.get(freeNode);
-      if (w != null) {
-        long current = w.getThreadIdToRunOn();
-        runnable.setThreadIdToRunOn(current);
+      for (T freeNode : freeNodes) {
+        IWorker<T> w = mapping.get(freeNode);
+        if (w != null) {
+          long current = w.getThreadIdToRunOn();
+          runnable.setThreadIdToRunOn(current);
+        }
+        mapping.put(freeNode, runnable);
       }
-      mapping.put(freeNode, runnable);
     }
   }
 
-  private void mapNodeToParent(List<T> freeNodes, int ix) {
+  private void mapNodeToParent(List<T> freeNodes) {
     if (!RuntimeBehavior.enforceThreadAffinity()) {
       return;
     }
-    T key = freeNodes.get(ix);
-    List<T> nodes = m_graph.getDependenciesFor(key);
-    nodes.forEach(eachNode -> upstream.put(eachNode, key));
+    for (T freeNode: freeNodes) {
```

Comment:
could be functionnal style with a transformation from free node to dependencie nodes

Reply:
Would be difficult because the inner `forEach()` iteration needs a reference to the outer freeNode iterating variable, which I think can't be done when you use a stream coupled with a transformation.